packages/store/src/actions/action-registry.ts

```diff
@@ -0,0 +1,117 @@
+import { inject, Injectable, OnDestroy, ɵisPromise } from '@angular/core';
+import {
+  defaultIfEmpty,
+  from,
+  isObservable,
+  mergeMap,
+  Observable,
+  of,
+  shareReplay,
+  takeUntil,
+  throwError
+} from 'rxjs';
+import { ɵActionHandlerMetaData } from '@ngxs/store/internals';
+
+import { ofActionDispatched } from '../operators/of-action';
+import { StateContextFactory } from '../internal/state-context-factory';
+import { ɵNgxsInternalDispatchedActions } from './dispatched-actions';
+
+// `action: Instance<ActionType>`.
+type ActionHandlerFn = (action: any) => void | Promise<void> | Observable<unknown>;
+
+interface InvokableActionHandlerMetaData extends ɵActionHandlerMetaData {
+  path: string;
+  instance: any;
+}
+
+@Injectable({ providedIn: 'root' })
+export class NgxsActionRegistry implements OnDestroy {
+  // Instead of going over the states list every time an action is dispatched,
+  // we are constructing a map of action types to lists of action metadata.
+  // If the `@@Init` action is handled in two different states, the action
+  // metadata list will contain two objects that have the state `instance` and
+  // method names to be used as action handlers (decorated with `@Action(InitState)`).
+  private readonly _actionTypeToHandlersMap = new Map<string, Set<ActionHandlerFn>>();
+
+  private readonly _stateContextFactory = inject(StateContextFactory);
+
+  private readonly _dispatched$ = inject(ɵNgxsInternalDispatchedActions);
+
+  ngOnDestroy(): void {
+    this._actionTypeToHandlersMap.clear();
+  }
+
+  get(type: string) {
+    return this._actionTypeToHandlersMap.get(type);
+  }
+
+  register(type: string, actionMeta: InvokableActionHandlerMetaData) {
+    const handlers = this._actionTypeToHandlersMap.get(type) ?? new Set();
+    const handler = this._createActionHandler(actionMeta);
+    handlers.add(handler);
+    this._actionTypeToHandlersMap.set(type, handlers);
+
+    return () => {
+      const handlers = this._actionTypeToHandlersMap.get(type)!;
+      handlers.delete(handler);
+    };
+  }
+
+  private _createActionHandler(actionMeta: InvokableActionHandlerMetaData) {
```

Comment:
As mentioned in the second comment of my previous review, the creation of the handler which in turn creates the state context, should stay outside of the action registry (in the StateFactory, for now).
Also the result handling logic should remain where it was in the StateFactory.